File: from-0901-to-1000/0931-minimum-falling-path-sum/solution.go
func minFallingPathSum(matrix [][]int) int {
    n := len(matrix)
    dp := make([][]int, n)
    for r := 0; r < n; r ++ {
        dp[r] = make([]int, n)
        copy(dp[r], matrix[r])
    }

    for r := 1; r < n; r ++ {
        for c := 0; c < n; c ++ {
            v := dp[r - 1][c]
            if c > 0 && dp[r - 1][c - 1] < v {
                v = dp[r - 1][c - 1]
            }
            if c < n - 1 && dp[r - 1][c + 1] < v {
                v = dp[r - 1][c + 1]
            }
            dp[r][c] += v
        }
    }

    result := dp[n - 1][0]
    for c := 1; c < n; c ++ {
        if result > dp[n-1][c] {
            result = dp[n - 1][c]
        }
    }
    return result
}

File: from-0901-to-1000/0931-minimum-falling-path-sum/solution.py
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        # Prepare DP matrix - just a copy of the matrix
        dp = [row.copy() for row in matrix]
        
        n = len(dp[0]) # the matrix is square
        for row in range(1, n):
            for col in range(n):
                # Add to the every cell in the current row
                # the min value from the 3 upper cells
                v = dp[row-1][col]
                if col > 0:
                    v = min(v, dp[row-1][col-1])
                if col < n-1:
                    v = min(v, dp[row-1][col+1])
                    
                dp[row][col] += v
        
        # The answer is the min value in the last row
        return min(dp[-1])
    